File: insight_gen.py
import os

from groq import Groq
from dotenv import load_dotenv

load_dotenv()
client = Groq(api_key=os.getenv("GROQ_API_KEY"))

_SYSTEM_PROMPT = "Write 2-3 sentence business insight. No bullets or headers. Use exact numbers. End with a recommendation."


def _format_prompt(user_query: str, result: dict) -> str:
    summary = result.get("summary", {})
    top_rows = result.get("data", [])[:3]
    metric = result.get("metric", "value")

    total = summary.get("total", "")
    max_label = summary.get("max_label", "")
    max_value = summary.get("max_value", "")

    rows_str = "; ".join(
        str({k: v for k, v in row.items()}) for row in top_rows
    )

    return (
        f"Q: {user_query}\n"
        f"Metric: {metric}. Total: {total}. Best: {max_label} ({max_value}).\n"
        f"Top rows: {rows_str}"
    )


def generate_insight(user_query: str, result: dict) -> str:
    if result.get("error"):
        return result.get("message", "An unknown error occurred.")

    if not result.get("data"):
        return "No data was returned for this query. Try adjusting your filters or broadening the date range."

    prompt = _format_prompt(user_query, result)

    try:
        response = client.chat.completions.create(
            model="llama-3.3-70b-versatile",
            messages=[
                {"role": "system", "content": _SYSTEM_PROMPT},
                {"role": "user", "content": prompt},
            ],
            temperature=0.7,
            max_tokens=200,
        )
        insight = response.choices[0].message.content.strip()
        if insight.startswith('"') and insight.endswith('"'):
            insight = insight[1:-1].strip()
        return insight
    except Exception:
        summary = result.get("summary", {})
        max_label = summary.get("max_label", "N/A")
        max_value = summary.get("max_value", 0)
        metric = result.get("metric", "value").replace("_", " ")
        return f"Top performer: {max_label} with ${max_value:,.0f} in {metric}."

File: frontend/src/lib/engine/insightGen.ts
/**
 * Insight Generator — TypeScript port of insight_gen.py
 * Generates 2-3 sentence AI insights from query results using Groq.
 */

import Groq from "groq-sdk";
import { QueryResult } from "./types";

const SYSTEM_PROMPT = "Write exactly 2-3 sentences. No bullets or headers. Use exact numbers from the data. End with one actionable recommendation.";

function formatPrompt(userQuery: string, result: QueryResult): string {
  const summary = result.summary ?? {};
  const topRows = (result.data ?? []).slice(0, 3);
  const metric = result.metric ?? "value";

  const rowsStr = topRows
    .map(row => JSON.stringify(Object.fromEntries(Object.entries(row).slice(0, 4))))
    .join("; ");

  return (
    `Q: ${userQuery}\n` +
    `Metric: ${metric}. Total: ${summary.total}. Best: ${summary.max_label} (${summary.max_value}).\n` +
    `Top rows: ${rowsStr}`
  );
}

function fallbackInsight(result: QueryResult): string {
  const summary = result.summary ?? {};
  const maxLabel = summary.max_label ?? "N/A";
  const maxValue = summary.max_value ?? 0;
  const rawMetric = result.metric ?? "value";
  const metric = rawMetric.replace(/_/g, " ");
  const prefix = /revenue|price|cost|amount/.test(rawMetric) ? "$" : "";
  return `Top performer: ${maxLabel} with ${prefix}${maxValue.toLocaleString(undefined, { maximumFractionDigits: 0 })} in ${metric}.`;
}

export async function generateInsight(userQuery: string, result: QueryResult): Promise<{ insight: string; fallback: boolean }> {
  if (result.error) {
    return { insight: result.message ?? "An unknown error occurred.", fallback: false };
  }
  if (!result.data?.length) {
    return { insight: "No data was returned for this query. Try adjusting your filters or broadening the date range.", fallback: false };
  }

  const apiKey = process.env.GROQ_API_KEY;
  if (!apiKey) {
    return { insight: fallbackInsight(result), fallback: true };
  }

  try {
    const client = new Groq({ apiKey });
    const prompt = formatPrompt(userQuery, result);

    const response = await client.chat.completions.create({
      model: "llama-3.3-70b-versatile",
      messages: [
        { role: "system", content: SYSTEM_PROMPT },
        { role: "user", content: prompt },
      ],
      temperature: 0.7,
      max_tokens: 200,
    });

    let insight = (response.choices[0].message.content ?? "").trim();
    if (insight.startsWith('"') && insight.endsWith('"')) {
      insight = insight.slice(1, -1).trim();
    }
    return { insight, fallback: false };
  } catch {
    return { insight: fallbackInsight(result), fallback: true };
  }
}
